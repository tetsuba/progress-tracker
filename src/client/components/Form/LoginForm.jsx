import React from 'react'
import { Button, Form, Col } from 'react-bootstrap'
import TextLink from '../TextLink/TextLink'
import { useInputChange } from '../../hooks/hooks'
import Box from '../Box/Box'

export default function LoginForm(props) {
  const { handleSubmit, error, hideLoginForm } = props
  const [inputs, setInputs] = useInputChange({
    email: 'test@test.com',
    password: '1234qwer',
  })

  return (
    <Box max={500}>
      <Col>
        <h3>Login</h3>
        <Form
          id="LoginForm"
          onSubmit={(e) => {
            e.preventDefault()
            const options = { variables: { input: inputs } }
            handleSubmit(options)
          }}
        >
          <Form.Group controlId="loginEmail">
            <Form.Label>Email address</Form.Label>
            <Form.Control
              required
              type="email"
              placeholder="Enter email"
              name="email"
              onChange={setInputs}
              value={inputs.email}
              isInvalid={!!error.email}
            />
            <Form.Control.Feedback type="invalid">
              {error.email && error.email.message}
            </Form.Control.Feedback>
          </Form.Group>

          <Form.Group controlId="loginPassword">
            <Form.Label>Password</Form.Label>
            <Form.Control
              required
              type="password"
              placeholder="Password"
              name="password"
              onChange={setInputs}
              value={inputs.password}
              isInvalid={!!error.password}
            />
            <Form.Control.Feedback type="invalid">
              {error.password && error.password.message}
            </Form.Control.Feedback>
          </Form.Group>
          <Button
            id="LoginFormSubmit"
            variant="primary"
            type="submit"
            className="float-right"
          >
            Submit
          </Button>
        </Form>
        <TextLink eventHandler={hideLoginForm}>Forgot password?</TextLink>
      </Col>
    </Box>
  )
}

LoginForm.defaultProps = {
  error: {
    email: undefined,
    password: undefined,
  },
}
