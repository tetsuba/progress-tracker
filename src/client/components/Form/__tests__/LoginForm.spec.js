import { testRenderer } from "../../../../test/testHelper";

// COMPONENTS
import LoginForm from "../LoginForm";


describe('<FormLogin>', () => {
    const baseProps = {
        handleSubmit: jest.fn(),
        resetPassword: jest.fn(),
    };

    describe('@Render', () => {
        it('should render the login form', () => {
            const props = {
                ...baseProps,
                error: {},
            };

            const wrapper = testRenderer(LoginForm, props);
            expect(wrapper).toMatchSnapshot()
        });

        it('should render the email verification form', () => {
            const props = {
                ...baseProps,
                error: { emailNotVerified: true },
            };

            const wrapper = testRenderer(LoginForm, props);
            expect(wrapper).toMatchSnapshot()
        });
    });

    describe('@Events', () => {
        const props = {
            ...baseProps,
            error: {},
        };

        it('should update email input with event onChange', () => {
            const wrapper = testRenderer(LoginForm, props);
            const email = { name: 'email' };
            const expected = 'test@test.com';

            wrapper.find(email).simulate('change', {
                persist: jest.fn(),
                target: {
                    name: 'email',
                    value: 'test@test.com',
                }
            });

            expect(wrapper.find(email).prop('value')).toEqual(expected)
        });

        it('should update password input with event onChange', () => {
            const wrapper = testRenderer(LoginForm, props);
            const password = { name: 'password' };
            const expected = '12345678';

            wrapper.find(password).simulate('change', {
                persist: jest.fn(),
                target: {
                    name: 'password',
                    value: '12345678',
                }
            });

            expect(wrapper.find(password).prop('value')).toEqual(expected)
        });

        it('should submit form with event onSubmit', () => {
            const wrapper = testRenderer(LoginForm, props);
            const id = '#loginForm';
            wrapper.find(id).simulate('submit');
            expect(props.handleSubmit).toHaveBeenCalledTimes(1)
        })
    });

    describe('@Error', () => {
        it('should render an error message if the error is an email or password error', () => {
            const props = {
                ...baseProps,
                error: {
                    email: {
                        message: 'This an email error',
                    },
                    password: {
                        message: 'This is a password error',
                    },
                },
            };
            const expected = Object.values(props.error)

            const wrapper = testRenderer(LoginForm, props);

            wrapper.find({type: 'invalid'}).forEach((message, i) => {
                expect(message.prop('children'))
                    .toEqual(expected[i].message)
            })
        });
    })
});
