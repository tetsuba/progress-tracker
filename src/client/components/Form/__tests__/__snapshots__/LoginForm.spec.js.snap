// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormLogin> @Render should render the email verification form 1`] = `
<Col>
  <EmailVerificationForm />
</Col>
`;

exports[`<FormLogin> @Render should render the login form 1`] = `
<Col>
  <h3>
    Login
  </h3>
  <Form
    id="loginForm"
    inline={false}
    onSubmit={[Function]}
  >
    <FormGroup
      controlId="loginEmail"
    >
      <FormLabel
        column={false}
        srOnly={false}
      >
        Email address
      </FormLabel>
      <FormControl
        isInvalid={false}
        name="email"
        onChange={[Function]}
        placeholder="Enter email"
        required={true}
        type="email"
        value="test@test.com"
      />
      <Feedback
        type="invalid"
      />
    </FormGroup>
    <FormGroup
      controlId="loginPassword"
    >
      <FormLabel
        column={false}
        srOnly={false}
      >
        Password
      </FormLabel>
      <FormControl
        isInvalid={false}
        name="password"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
        value="1234qwer"
      />
      <Feedback
        type="invalid"
      />
    </FormGroup>
    <Button
      active={false}
      className="float-right"
      disabled={false}
      id="loginButton"
      type="submit"
      variant="primary"
    >
      Submit
    </Button>
  </Form>
  <TextLink
    eventHandler={[MockFunction]}
  >
    Forgot password?
  </TextLink>
</Col>
`;
